src/compiler/checker.ts

```diff
@@ -6163,6 +6167,12 @@ namespace ts {
             return !!(typeParameter.symbol && forEach(typeParameter.symbol.declarations, decl => isTypeParameterDeclaration(decl) && decl.default));
         }
 
+        function getConstraintType(type: Type): Type {

```

Comment:
There is a thicket of related concepts here now that need better names: base constraint, syntactic constraint and this concept — constraint? — that I think is the correct one to use most of the time. But it doesn’t support index access types, for example. 

Let’s talk in person and look over the current set of constraint-related functions. I don’t think we should be using the syntactic constraint much anymore, but I’m not sure of the best way to make that happen.

Reply:
`TypeFlags.TypeVariable` includes indexed accesses.

Reply:
After talking in person, I've since swapped just reusing mostof `getBaseConstraintOfType`, since that handles most of the behaviors we wanted anyway (including handling intersections correctly).